linkerd/app/core/src/retry.rs

```diff
@@ -128,10 +132,60 @@ impl<B: Default + HttpBody> CloneRequest<http::Request<B>> for () {
         *clone.headers_mut() = req.headers().clone();
         *clone.version_mut() = req.version();
 
-        // // Count retries toward the request's total handle time.
-        // if let Some(ext) = req.extensions().get::<handle_time::Tracker>() {
-        //     clone.extensions_mut().insert(ext.clone());
-        // }
+        Some(clone)
+    }
+}
+
+// === impl WithBody ===
+
+impl WithBody {
+    /// Allow buffering requests up to 64 kb
+    pub const MAX_BUFFERED_BYTES: usize = 64 * 1024;
+}
+
+impl<B: HttpBody + Unpin + Default> CloneRequest<http::Request<buf_body::BufBody<B>>> for WithBody {
+    fn clone_request(
+        req: &http::Request<buf_body::BufBody<B>>,
+    ) -> Option<http::Request<buf_body::BufBody<B>>> {
+        let content_length = |req: &http::Request<_>| {
+            req.headers()
+                .get(http::header::CONTENT_LENGTH)
+                .and_then(|value| value.to_str().ok()?.parse::<usize>().ok())
+        };
+
+        // Requests without bodies can always be retried, as we will not need to
+        // buffer the body. If the request *does* have a body, retry it if and
+        // only if:
+        // - the request is a POST
+        // - the request contains a `content-length` header,
+        // - the content length is >= 64 kb,
+        let has_body = !req.body().is_end_stream();
+        let method = req.method();
+        if has_body
+            && (method != http::Method::POST
```

Comment:
Do we actually care about the method? Why not support PUTs etc as well? We should really just let this configuration come from profiles

Reply:
linkerd/linkerd2#6130 mentions POSTs by name, so i (incorrectly) assumed that we didn't want to retry any other methods. i'll change it.

Reply:
incidentally, we could certainly also make the maximum content-length to buffer be configurable as well, but i thought that seemed like a follow-up --- we could either configure it on the profile or globally with an annotation?

Reply:
> incidentally, we could certainly also make the maximum content-length to buffer be configurable as well, but i thought that seemed like a follow-up --- we could either configure it on the profile or globally with an annotation?
 
Yeah, I think it can be configurable. It seems like a process-specific thing more than a service-specific setting, so I'd probably start with an environment configuration.

Reply:
removed the POST requirement in 25e996e95caba4ac277da61ac8cc1f310c986f4b; we can make the max buffer size configurable in a follow-up?